Matter & Memfault (#325)
memfault/interrupt: _drafts/memfault-matter.md

At Tridonic, we see Matter as the force unifying the Smart Home and eventually the professional lighting space as well, thus consolidating the wireless options and giving new opportunities to use the consolidated wireless eco-system for the benefit of the users. Thread has been chosen for the suitability to support low-power sensor devices and due to its nature for handling multicast efficiently to a multitude of wireless devices.

The Tridonic products are therefore based on matter/Thread/802.15.4. Thus I'll focus on UDP, IPv6, and Thread in the article.

## Methods to connect a Matter/Thread device to Memfault

There is a lot of flexibility that comes with IP. This gives us multiple options to send data to Memfault.

1. Directly via HTTP & TCP
    1. via IPv6
    2. via IPv4
2. via a Memfault UDP relay
    1. via a public Memfault UDP relay
        1. DNS entry for a Memfault UDP relay <memfault-udp-relay.example.com> (if required via NAT64)
    2. via a local Memfault UDP relay
        1. via an announcement using multicast DNS service discovery

>         2. via a well-known IPv6 multicast address, e.g.<ff05::f417> (shortened hexspeak for Memfault)

Author: it's a shortened and hex converted / leetspeak (f417 -> memFAuLT).

>         2. via a well-known IPv6 multicast address, e.g.<ff05::f417> (shortened hexspeak for Memfault)

Author: f -> F
4 -> A
1 -> L
7 -> T

>         2. via a well-known IPv6 multicast address, e.g.<ff05::f417> (shortened hexspeak for Memfault)

Author: one could use any address, but that was easy to use and would a minor hint for the purpose of the multicast packets when looking at network traces.

In the following, we'll discuss the various options.

### Via HTTP & TCP

TCP is currently an option for non-Thread Matter devices. TCP is being discussed and implemented also for Thread but might have an impact on the code size.

This could be enabled with

```kconfig
CONFIG=OPENTHREAD_CONFIG_TCP_ENABLE=y
CONFIG_HTTP_CLIENT=y

> ```

Author: done

---
title: Matter, Thread, and Memfault
description:
  Matter and Thread are changing the smart home landscape with IPv6-based
  connectivity. This post shows how to use Memfault in a matter/Thread
  environment - the changes required to the Memfault SDK, the required set-up
  and ideas for future integration of Memfault with Thread.
author: mabe
image: img/<post-slug>/cover.png # 1200x630
---

<!-- excerpt start -->

> I"m Markus, software engineer @ [Tridonic](https://tridonic.com), where we are working on [Internet-connected wireless lighting solutions](http://www.tridonic.com/matter) based on the [matter standard](https://csa-iot.org/all-solutions/matter/). To be able to monitor the reliability of those devices we've been using Memfault and tied it into matter/Thread and its UDP/IPv6 stack based on the Nordic Connect SDK. In the following, I'll show you the modifications we've done to enable Memfault in an IPv6 solution.

Reviewer: ```suggestion
I'm Markus, software engineer @ [Tridonic](https://tridonic.com), where we are working on [Internet-connected wireless lighting solutions](http://www.tridonic.com/matter) based on the [matter standard](https://csa-iot.org/all-solutions/matter/). To be able to monitor the reliability of those devices we've been using Memfault and tied it into matter/Thread and its UDP/IPv6 stack based on the Nordic Connect SDK. In the following, I'll show you the modifications we've done to enable Memfault in an IPv6 solution.
```

<!-- excerpt end -->

{% include newsletter.html %}

{% include toc.html %}

## matter & Thread

> matter <https://csa-iot.org/all-solutions/matter/> is an interoperability standard for the IoT from the [connectivity standards alliance (csa)](https://csa-iot.org) previously known as the zigbeeAlliance. The matter standard is implemented in the open-source reference implementation in the [GitHub project-chip repository](https://github.com/project-chip/connectedhomeip/). matter currently uses UDP on top of IP(v6) and can run on top of different lower layers: Ethernet, WiFi, Thread/802.15.4. [Thread](https://www.threadgroup.org/) is a low-power IPv6 connectivity standard. Its reference implementation is available in the [openThread Github repository](https://github.com/openthread/openthread). The matter SDK as well as the openThread SDK are also available bundled in the [Nordic Connect SDK](https://developer.nordicsemi.com/nRF_Connect_SDK/).

Reviewer: Can you capitalize `Matter` throughout the article? I see that the stylized logo is lower case, but everywhere on there site when mentioned in text it is capitalized.

Also here I think `Zigbee Alliance` needs to be capitalized and two words.

Author: done

> matter <https://csa-iot.org/all-solutions/matter/> is an interoperability standard for the IoT from the [connectivity standards alliance (csa)](https://csa-iot.org) previously known as the zigbeeAlliance. The matter standard is implemented in the open-source reference implementation in the [GitHub project-chip repository](https://github.com/project-chip/connectedhomeip/). matter currently uses UDP on top of IP(v6) and can run on top of different lower layers: Ethernet, WiFi, Thread/802.15.4. [Thread](https://www.threadgroup.org/) is a low-power IPv6 connectivity standard. Its reference implementation is available in the [openThread Github repository](https://github.com/openthread/openthread). The matter SDK as well as the openThread SDK are also available bundled in the [Nordic Connect SDK](https://developer.nordicsemi.com/nRF_Connect_SDK/).

Reviewer: ```suggestion
[matter](https://csa-iot.org/all-solutions/matter/) is an interoperability standard for IoT devices from the [connectivity standards alliance (csa)](https://csa-iot.org) previously known as the zigbeeAlliance. The matter standard is implemented in the open-source reference implementation in the [GitHub project-chip repository](https://github.com/project-chip/connectedhomeip/). matter currently uses UDP on top of IP(v6) and can run on top of different lower layers: Ethernet, WiFi, Thread/802.15.4. [Thread](https://www.threadgroup.org/) is a low-power IPv6 connectivity standard. Its reference implementation is available in the [openThread Github repository](https://github.com/openthread/openthread). The matter SDK as well as the openThread SDK are also available bundled in the [Nordic Connect SDK](https://developer.nordicsemi.com/nRF_Connect_SDK/).
```

<!--

TODO: more on matter and Thread.....

-->

> The Tridonic products are based on Thread/802.15.4. Thus I'll focus on UDP, IPv6, and Thread in the following.

Reviewer: ```suggestion
The Tridonic products are based on Thread/802.15.4. Thus I'll focus on UDP, IPv6, and Thread in the article.
```

## Methods to connect a matter/Thread device to Memfault

> There"s a lot of flexibility that comes with IP and thus multiple options to send data to Memfault.

Reviewer: ```suggestion
There is a lot of flexibility that comes with IP. This gives us multiple options to send data to Memfault.

#### Via public UDP relay via IPv6

If the UDP relay would be reachable via IPv6 one could use

```shell
> dns resolve memfaultudprelay.example.com
> udp send <IPv6 address> 20001 hello
```

#### Via public UDP relay via IPv4/NAT64

If the UDP relay is only reachable via IPv4:

```shell
> dns resolve4 memfaultudprelay.example.com
> udp send <NAT64 address> 20001 hello
```

#### Via mDNS

Another mechanism could be to distribute this service to the devices via multicast DNS service discovery which is also used by matter devices.

E.g. on a machine one could advertise a Memfault service:

```shell
dns-sd -R MEMFUDP _memfault._udp . 20001
# or
avahi-publish-service ...
avahi-publish -s MEMFUDP _memfault._udp 20001
```

Or announce it via the Border Router:

```shell
> srp client host name BR_MFLT_UDP
> srp client host address ML-EID
> srp client service add otbr1 _memfault._udp 20001
```

This could be tested on the device:

```shell
> dns browse _memfault._udp.default.service.arpa
> udp send <HostAddress> 20001 hello
```

#### Via multicast

> We have not been using and implementing any of the options from above but instead decided to send the memfault reports to a dedicated IPv6 multicast address. The UDP relay is listening on IPv6 and joins that multicast group.

Reviewer: Capitalize Memfault throughout

Author: done

This mechanism requires very little setup on the customer premise and very implementation in the device. If the customer has no UDP relay, no data is transferred - which is good from a privacy point of view. However, if a customer has a complaint error reporting can easily be enabled.

On the device shell, this could be tested with:

```shell
> udp send ff05::f417 20001 hello
```

This approach is described in more detail in the following sections.

## Getting to a working matter/Memfault solution

Memfault offers a [UDP relay sample](https://github.com/memfault/memfault-nRF9160-relay/tree/main/server). We'll use this relay and adapt it.

Additionally, we'll change the default Memfault MCU SDK (respectively Nordic's integration of it) and adapt it to IPv6.

### Firmware

> Nordic provides a [matter light_bulb sample](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/matter/light_bulb/README.html) as part of their SDK which we'll use as a base for this investigation and compile it according to the provided instructions.

Reviewer: ```suggestion
Nordic provides a [matter light bulb sample](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/matter/light_bulb/README.html) as part of their SDK which we'll use as a base for this investigation and compile it according to the provided instructions.
```

> Following [Memfault's MCU nRF Connect SDK Guide](https://docs.memfault.com/docs/mcu/nrf-connect-sdk-guide/), we'll take the example code and combine it with the matter light_bulb sample.

Reviewer: ```suggestion
Following [Memfault's MCU nRF Connect SDK Guide](https://docs.memfault.com/docs/mcu/nrf-connect-sdk-guide/), we'll take the example code and combine it with the matter light bulb sample.
```

* The script has been changed to use an IPv6 socket.
* The script joins the IPv6 multicast group on the relevant interface (here `utun3`, adapt to your interface name).

```python
INTERFACE = 'utun3'
LOCAL_PORT = 20001
BUFFER_SIZE = 508 + 28

MEMFAULT_IPV6_GROUP_6 = 'ff05::f417'

addrinfo = socket.getaddrinfo(MEMFAULT_IPV6_GROUP_6, None)[0]

print(addrinfo)

UDPServerSocket = socket.socket(addrinfo[0], type=socket.SOCK_DGRAM)
UDPServerSocket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
UDPServerSocket.bind(('', LOCAL_PORT))

group_bin = socket.inet_pton(addrinfo[0], addrinfo[4][0])
mreq = group_bin + struct.pack('@I', socket.if_nametoindex(INTERFACE))
UDPServerSocket.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_JOIN_GROUP, mreq)

print(f"Listening on interface {INTERFACE} port {LOCAL_PORT}")
```

### Build the light_bulb sample

You can build the light_bulb sample according to [Nordic's documentation](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/2.3.0/nrf/samples/matter/light_bulb/README.html).

## Run Border Router and Commission matter device

> Set up a border router and commission the matter light_bulb according to the [Nordic matter documentation](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/protocols/matter/index.html).

Reviewer: Change all non-links to `light bulb`

Author: done
